sys/kerncore/src/lib.rs

```diff
@@ -159,35 +178,53 @@ where
 
     // Per the function's preconditions, the region table is sorted in ascending
     // order of base address, and the regions within it do not overlap. This
-    // lets us use a one-pass algorithm.
+    // lets us use a binary search followed by a short scan
     let mut scan_addr = slice.base_addr();
     let end_addr = slice.end_addr();
 
-    for region in table {
-        if region.contains(scan_addr) {
-            // Make sure it's permissible!
-            if !region_ok(region) {
-                // bail to the fail handling code at the end.
-                break;
-            }
-
-            if end_addr <= region.end_addr() {
-                // We've exhausted the slice in this region, we don't have
-                // to continue processing.
-                return true;
-            }
+    let Ok(index) =
+        table.binary_search_by(|reg| region_compare(reg, scan_addr))
+    else {
+        // No region contained the start address.
+        return false;
+    };
+
+    // Perform fast checks on the initial region. In practical testing this
+    // provides a ~1% performance improvement over only using the loop below.
+    let first_region = &table[index];
+    if !region_ok(first_region) {
+        return false;
+    }
+    // Advance to the end of the first region
+    scan_addr = first_region.end_addr();
+    if scan_addr >= end_addr {
+        // That was easy
```

Comment:
a super tiny, unimportant copyedit nit that i feel quite silly for commenting on: all the other comments in this function are terminated by periods, but these aren't (exclamation point optional):
```suggestion
    // Advance to the end of the first region.
    scan_addr = first_region.end_addr();
    if scan_addr >= end_addr {
        // That was easy!
```

Reply:
¡but what if I want it at the front though